### PR TITLE
Infer iOS-specific build options 

### LINKED ISSUE #1089: Fix resolution of iOS specific values
Discovered while investigating and fixing #1067.

For Flutter to build a complete iOS apk, only a flavor is needed:

```console
flutter build ios -t integration_test/daily_mission_test.dart --flavor dev
```

But for Patrol, the user has to specify `configuration` and `scheme` explicitly:

```console
patrol develop -t integration_test/daily_mission_test.dart --flavor dev --configuration Debug-dev --scheme dev
```

Let's make `patrol` work like `flutter`.

### PR DESCRIPTION
**Pull request by bartekpacia**
Fixes #1089 

### Example

Case: the user wants to run tests of the app with flavor `dev` on iOS.

**Before this PR**

```console
patrol test --flavor dev --scheme dev --configuration Debug-dev
```

**After this PR**

iOS-specific build options are inferred.

```console
patrol test --flavor dev
```

### CODE REVIEW

**Review by shilangyu — APPROVED** [review]

File: packages/patrol_cli/lib/src/crossplatform/app_options.dart [hunk 1]
```diff
@@ -135,7 +136,7 @@ class IOSAppOptions {
 
   /// Translates these options into a proper flutter build invocation, which
   /// runs before xcodebuild and performs configuration.
-  List<String> toFlutterBuildInvocation() {
+  List<String> toFlutterBuildInvocation(BuildMode buildMode) {
```
Comment (Albert221) [inline packages/patrol_cli/lib/src/crossplatform/app_options.dart:139]: Do you accept `buildMode` for the future? You don't use it now

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart [hunk 2]
```diff
@@ -13,6 +13,38 @@ import 'package:patrol_cli/src/devices.dart';
 import 'package:patrol_cli/src/ios/ios_deploy.dart';
 import 'package:process/process.dart';
 
+enum BuildMode {
+  debug('Debug'),
+  profile('Profile'),
+  release('Release');
+
+  const BuildMode(this.name);
```
Comment (Albert221) [inline packages/patrol_cli/lib/src/ios/ios_test_backend.dart:21]: Do you ever intend on using this constructor? If not, make it private

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart [hunk 3]
```diff
@@ -13,6 +13,38 @@ import 'package:patrol_cli/src/devices.dart';
...
+    return values.firstWhere((value) => value.name == name);
+  }
+
+  final String name;
+
+  static const _defaultScheme = 'Runner';
+
+  String createScheme(String? flavor) {
+    if (flavor == null) {
+      return BuildMode._defaultScheme;
```
Comment (Albert221) [inline packages/patrol_cli/lib/src/ios/ios_test_backend.dart:35]: Isn't the `BuildMode.` part redundant?

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart [hunk 4]
```diff
@@ -149,8 +181,6 @@ class IOSTestBackend {
   }
 
   Future<void> uninstall(String appId, Device device) async {
-    _logger.info('Uninstalling $appId from ${device.name}');
```
Comment (Albert221) [inline packages/patrol_cli/lib/src/ios/ios_test_backend.dart:152]: Why? :/

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart [hunk 5]
```diff
@@ -227,8 +251,8 @@ class IOSTestBackend {
   }
 
   Future<String> _sdkVersion(bool real) async {
-    // See the versions yourself:
-    // $ xcodebuild -showsdks -json | jq ".[] | {sdkVersion, platform}"
+    // See the versions yourself: $ xcodebuild -showsdks -json | jq ".[] |
+    // {sdkVersion, platform}"
```
Comment (Albert221) [inline packages/patrol_cli/lib/src/ios/ios_test_backend.dart:255]: You just made it hard to copy. Nothing wrong in long comments like that

File: packages/patrol_cli/lib/src/runner/patrol_command.dart [hunk 6]
```diff
@@ -36,6 +37,24 @@ abstract class PatrolCommand extends Command<int> {
...
+        help: 'Build a debug version of your app',
+        defaultsTo: true,
+      )
+      ..addFlag(
+        'profile',
+        help: 'Build a version of your app for performance profiling.',
+      )
+      ..addFlag(
+        'release',
+        help: 'Build a release version of your app (default mode)',
```
Comment (Albert221) [inline packages/patrol_cli/lib/src/runner/patrol_command.dart:54]: But `debug` has `defaultsTo: true` o_O

File: packages/patrol_cli/lib/src/runner/patrol_command.dart [hunk 7]
```diff
@@ -146,4 +149,22 @@ abstract class PatrolCommand extends Command<int> {
   List<String> stringsArg(String name) {
     return argResults![name]! as List<String>? ?? <String>[];
   }
+
+  BuildMode get buildMode {
+    if (!_usesBuildOption) {
+      throwToolExit('This command does not support build mode option');
```
Comment (Albert221) [inline packages/patrol_cli/lib/src/runner/patrol_command.dart:155]: Do you explain that `debug`, `profile` and `release` are actually _build modes_? If not, this message may still be ambiguous

File: packages/patrol_cli/test/ios/ios_test_backend_test.dart [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        const buildMode = BuildMode.debug;
+        const flavor = 'dev';
+
+        expect(buildMode.createConfiguration(flavor), 'Debug-dev');
+        expect(buildMode.createScheme(flavor), 'dev');
+      });
+
+      test('infers build options in release mode when flavor is null', () {
+        const buildMode = BuildMode.release;
+        String? flavor;
```
Comment (Albert221) [inline packages/patrol_cli/test/ios/ios_test_backend_test.dart:25]: I'd assign `null` to it explicitly and make it a const and disable the lint for default values for this file or for those lines

**Review by Albert221 — CHANGES_REQUESTED** [review]

File: packages/patrol_cli/lib/src/crossplatform/app_options.dart (hunk 1 above)
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/crossplatform/app_options.dart:139]: thanks, fixed

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart (hunk 2 above)
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/ios/ios_test_backend.dart]: It's not needed anymore, removed. Thanks.

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart (hunk 3 above)
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/ios/ios_test_backend.dart:34]: It is, hah. Thanks.

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart (hunk 4 above)
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/ios/ios_test_backend.dart:152]: If `--verbose` flag is passed to `patrol_cli`, the executed subprocesses (in this case, app uninstallation commands) are logged automatically by [LoggingLocalProcessManager from here](https://github.com/leancodepl/patrol/blob/master/packages/patrol_cli/lib/src/base/process.dart).

So, these lines I removed were unnecessary duplicates.

File: packages/patrol_cli/lib/src/runner/patrol_command.dart (hunk 6 above)
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/runner/patrol_command.dart:54]: oopsie, copied from `flutter build ios` and forgot to change

File: packages/patrol_cli/test/ios/ios_test_backend_test.dart (hunk 8 above)
Comment (bartekpacia) [reply packages/patrol_cli/test/ios/ios_test_backend_test.dart:25]: Nice but I don't agree. 

> Special cases aren't special enough to break the rules.
>
> ~ The Zen of Python

File: packages/patrol_cli/lib/src/runner/patrol_command.dart (hunk 7 above)
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/runner/patrol_command.dart:155]: This is a classic example of "should never happen".

Thanks for spotting this. I added the following comment:

> // If this happens, it's a developer fault, not the user's.

I guess the future me will thank the present me :D

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart (hunk 5 above)
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/ios/ios_test_backend.dart]: Right. FIxed as well.

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart (hunk 2 above)
Comment (shilangyu) [reply packages/patrol_cli/lib/src/ios/ios_test_backend.dart]: > Do you ever intend on using this constructor? If not, make it private
It is an enum, you cannot construct a variant dynamically

File: packages/patrol_cli/test/ios/ios_test_backend_test.dart (hunk 8 above)
Comment (Albert221) [reply packages/patrol_cli/test/ios/ios_test_backend_test.dart:25]: I don't agree with your not agreeing

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart (hunk 2 above)
Comment (Albert221) [reply packages/patrol_cli/lib/src/ios/ios_test_backend.dart]: @shilangyu but you may use it for enum-value which is what I asked for

**Review by Albert221 — APPROVED** [review]

**Pull request merged**